### DISCUSSION TITLE
Subprocess / child process management in kj?

### DISCUSSION THREAD

**Discussion by jhasse** [capnproto/capnproto#1966] [Ideas]
Hi,

I was wondering about using Cap'n'Proto for a version of [ninja](https://github.com/ninja-build/ninja) which uses a frontend and a backend process. There's already a protobuf implementation for this with custom IPC, but I was brainstorming about using Cap'n'Proto for this and fully using kj's main loop.
The main job of a build tool is spawning subprocesses and reading their output, if that could be integrated into kj's epoll main loop that would be great. Is there any functionality that could help me in that regard? Do you think such functionality would be accepted upstream or is it out of scope?

**Comment by kentonv** [comment]
There is actually some support for subprocesses already, via `UnixEventPort::onChildExit()`:

https://github.com/capnproto/capnproto/blob/eef235dfa7aa40087605538ee3942b75a8910787/c%2B%2B/src/kj/async-unix.h#L138

This is sufficient for Unix-based apps to wait on child processes in the KJ event loop without modifying KJ.

However:
* There is no support for waiting on subprocesses in Windows. The Windows EventPort implementation is based on I/O Completion Ports which, to everyone's great frustration, do not support waiting for every kind of event in the Win32 API. We would need an alternate EventPort implementation based on MsgWaitForMultipleObjectsEx() instead. This is something I always intended to have, but Windows has not been a high priority. If someone wants to build this I'm potentially interested in merging it. (BTW there is a secret internal NT API which allows I/O completion to be delivered via APC, which actually makes MsgWaitForMultipleObjectsEx() a strict superset of IOCP for single-threaded event loops like KJ's. The ideal KJ Win32 event loop would probably be based on that.)
* It would be nice to present a higher-level, cross-platform API for subprocesses. However, I don't know how realistic this really is. Even across different favors of Unix, the available functionality differs drastically (e.g. setting up namespaces on Linux). And on Windows, as we all know, subprocesses are completely different in every possible way. For this reason, I've found myself writing one-off subprocess APIs as part of various downstream projects ([Sandstorm](https://github.com/sandstorm-io/sandstorm/blob/63f265d815bd511a5250927692adfe59e317d500/src/sandstorm/util.h#L266), Cloudflare Workers) designed to address only specific use cases, but I've never felt what I came up with was worthy of being part of KJ itself. I'd be happy for contributions here _if_ we can agree on what functionality it should have and what the API should look like. But it might be easier to just write code downstream, like I have.

**Reply by alexshpilkin** [reply]
@kentonv wrote:
> BTW there is a secret internal NT API which allows I/O completion to be delivered via APC, which actually makes MsgWaitForMultipleObjectsEx() a strict superset of IOCP for single-threaded event loops like KJ's.

The documentation for `WaitForMultipleObjects` et al. says that an alertable invocation can be interrupted by either a user APC _or an I/O completion routine._ Seems like what you’re proposing would just be reimplementing I/O completion routines (“alertable I/O”) on top of APCs the same way Windows itself does. Are there any I/O actions that can be performed with an IOCP but not with a completion routine?

([Richter](https://www.microsoftpressstore.com/store/windows-via-c-c-plus-plus-9780735639218) says alertable I/O is “horrible”, but then his biggest actual gripe against it is bad interaction with multithreading—that a thread that issued an I/O has to be the one waiting for it. You specifically mention single-threaded use cases.)

**Reply by kentonv** [reply]
Hmm. Sorry, I'm actually going off my own tweet from 2017: https://twitter.com/KentonVarda/status/937856872496762880

I cannot remember what it is that I found that I did not believe existed outside of the internal APIs. Maybe it was actually I/O completion routines and somehow I didn't realize they were already exposed publicly? From a brief look it does seem like I/O completion routines accomplish what we want, but I remember having read all these docs carefully in the past and concluding that there was no all-encompassing solution... did I just miss it?

**Reply by kentonv** [reply]
I must have just missed this. I don't understand how. Anyway, clearly KJ's Win32 event port should be rewritten using WaitForMultipleObjectsEx and completion routines, since it's a single-threaded event loop anyway.
